### ISSUE TITLE
Create visually customizable keyring format

### ISSUE THREAD

**Issue by danfinlay** [MetaMask/metamask-extension#3249]
Opening for discussion, to hone the approach and criteria for the goal of allowing arbitrary signers to be added to MetaMask.

Extending from #328, but adding a visual component. The related design issue is #731. This could be done in parallel with #731 as long as it was kept design-agnostic.

Blocking #717, #605, #716.

# MetaMask Mutli-Signer Acceptance Criteria

An Aspect of MetaMask that has waited far too long is our ability to support multiple signing strategies. At time of writing, MetaMask only accepts two types of signer (both key pairs), the 12-word mnemonic, and the imported private key.

Going forward, MetaMask strives to allow users to sign blockchain transactions representing any of a multitude of strategies, including:

- Hardware Wallets (Ledger, Trezor, Firefly)
- Offline Signers (copy the params, paste the signed tx, or all QR codes?)
- Contract Based Accounts (Multi Sig, uPort, 2FA, etc)
- Remote/Hosted signers (Imagine if Coinbase signed as a service)

## The Keyring Class

Our initial effort towards this end was defining [The Keyring Class](https://github.com/MetaMask/eth-simple-keyring#the-keyring-class-protocol), a JavaScript protocol that objects can adhere to, and be added to MetaMask as new signing strategies.

What we never added to the Keyring protocol was a notion of custom signing views, which would definitely be necessary for things like offline signers or contract based accounts, since these might have options, preferences, and parameters that only relate to that view itself.

## The Work To Go

This work would be composed of several sub-features, which should stem from our New UI work, which currently lives on the `uat` branch (hopefully to merge this week and deploy next).

A complete implementation of this platform would be to implement a single minimum-viable new signing strategy (like the Ledger, or an offline signer with custom UI). The exact signer would not be as important as the platform's easy extensibility.

The roadmap would roughly include:

- The account list should group by Keyring, with headers.
- The `import account` view should list a new signer type.
- That new signer import screen should include whatever that signer requires to set up. (Opening in new tab is probably a good idea for a complex form).
- When an account from that signer is selected, and a transaction is proposed, this signer would get to provide its own confirmation screen/view.

## Confirmation View Considerations

The confirmation view itself is not perfectly defined. Currently our confirmation screen strives to provide all info that the user needs to approve a signature, and no other action is required, since MetaMask holds those keys.

For something like a hardware wallet, it is easy to imagine that a subsequent screen or series of steps could be provided after this confirmation screen, and this might be acceptable for only adding Ledger support, but other signers have other requirements!

Consider a multi-sig or contract account. The confirmation screen in this scenario would need to refer to either:
- The transaction the user intends to send
- The transaction to their contract account, which will then proxy that other transaction.

Since the former is the one that is intelliglbe to the user, we might prefer to show this by default, but since the latter would show the gas parameters that are actually relevant to the user, it is clear that the actual confirmation screen for a contract account would need to be some kind of fusion.

For this reason, custom confirmation screens and flows probably need to be (at least an optional) customization provided by new Keychains.

## Signer Composability

The last major criteria, and some might argue whether it is truly a requirement, would be signer composability. This is the notion that some signers might require another signer to sign, and so MetaMask could empower these signers greatly by providing the user options of how to compose their signers.

For example, in the case of a multi-sig wallet, a user might add a contract account which will forward any transaction which includes a signature from two of its authorized keys. The user might hold one key on their phone, and one key on their computer.

This detail could be left to the custom signer, and as a minimum viable product, that might be what we agree to do. However, what if the user wants their local key to be a hardware wallet? In this case, the user wants to use a contract account, which might impose a spending limit, and they only want to keep their keys on a hardware device. I don't think we should expect each contract wallet keyring author to implement their own hardware wallet support, and so MetaMask could offer some powerful leverage for contract wallets by allowing signers to be arbitrarily composed.

In the most complicated case, this might mean a hardware wallet is authorized to sign fot a contract identity which is authorized to sign for a multi-sig account.

I know, this may be premature optimization, and we can definitely ship smaller pieces sooner, but I think keeping a consideration for this long-term path could help our architecture stay flexible going forward, because I think signer composability is eventually going to be an obviously desired feature; People will want to browse web3 sites on behalf of organizations that they are part of, and they will want to propose transactions to that organization, and they will propose those transactions from their uPort identities (contract accounts), which they will likely authorize from their phones or hardware devices.

## Acceptable Formats

Ideally, Keyrings themselves can be defined within a specified bundle or module, paving the way for a "signer store", so that eventually we can further abstract the job of selecting signers away from the core codebase to whatever mechanism we agree is the most secure and usable.

## Bounty Warnings

If you pursue a bounty on this issue, to be clear: While you pursue it, the issue is not reserved for you. We will pursue this issue as soon as possible, and the only way you will earn a bounty for this issue is if you both deliver this feature before we're able to, and it satisfies our very high standards, which we will enforce rigorously.

Pursue at your own risk!

**Comment by danfinlay** [comment]
## Sandboxed execution

Especially if we were making a signer store, or trying to accept signing strategies more casually, isolating new signers, either through their UI in an iframe, or through their background process in a web-worker, would be an important set of considerations.

**Comment by danfinlay** [comment]
We're considering posting a bounty on this, as a challenge to the very rambunctious and sometimes impatient crowd waiting for these valuable features.

If you pursue a bounty on this issue, to be clear: While you pursue it, the issue is not reserved for you. We will pursue this issue as soon as possible, and the only way you will earn a bounty for this issue is if you both deliver this feature before we're able to, and it satisfies our very high standards, which we will enforce rigorously. Pursue at your own risk!

**Comment by gitcoinbot** [comment]
__This issue now has a funding of 1.1 ETH (1010.24 USD) attached to it.__

 * If you would like to work on this issue you can claim it [here](https://gitcoin.co/funding/details?url=https://github.com/MetaMask/metamask-extension/issues/3249).
 * If you've completed this issue and want to claim the bounty you can do so [here](https://gitcoin.co/funding/details?url=https://github.com/MetaMask/metamask-extension/issues/3249)
 * Questions? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $9560.46 more Funded OSS Work Available at: https://gitcoin.co/explorer

**Comment by lazaridiscom** [comment]
I have somehow the impression that just the necessary pre-work (Requirements Analysis, Isolating Subsystems and Tasks etc.) would eat up this bounty. It feels like an whole "sub-project" within the MetaMask project. But can be wrong.

The problem I (and most possibly others) have, is that I do not understand the problem, thus I'm unable to estimate the effort.

## I know this:

* MetaMask cannot use e.g. Trezor to sign transactions
* Users want this much, the market wants this much.
* Team very busy, Issue-count raises and raises

and this:

* Task A: a flexible sign mechanism should be provided
* Task B: a UI using this sign mechanism should be provided

is this valid? Or have I lost it already?

**Comment by danfinlay** [comment]
@lazaridiscom You're completely right, both about this feature being sub-project sized, and our team being very busy 😬.

I definitely think this bounty is currently insufficient, and I think we will gradually add more to it (and encourage others to add to it), but we also are budgeted to complete this internally (and we are hiring to help on it), so it's a weird issue for us to add bounties on.

If you're seriously interested in contributing more (it sounds like you'd be a great help), please reach out, maybe via email or something. A sub-project this large is hard to treat as a simple bounty, but with enough coordination, we could probably work something out.

**Comment by lazaridiscom** [comment]
Send you and email.

**Comment by gitcoinbot** [comment]
__Work has been started on the 1.1 ETH (980.04 USD) funding by__: 
 1. [@lazaridiscom](https://gitcoin.co/profile/lazaridiscom) 

 __Please work together__ and coordinate delivery of the issue scope. Gitcoin doesn't know enough about everyones skillsets / free time to say who should work on what, but we trust that the community is smart and well-intentioned enough to work together.  As a general rule; if you start work first, youll be at the top of the above list ^^, and should have 'dibs' as long as you follow through. 

 On the above list? Please leave a comment to let the funder (@owocki) and the other parties involved what you're working, with respect to this issue and your plans to resolve it.  If you don't leave a comment, the funder may expire your submission at their discretion. 

 * Learn more [on the gitcoin issue page](https://gitcoin.co/funding/details?url=https://github.com/MetaMask/metamask-extension/issues/3249)
 * Questions? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $12415.8 more Funded OSS Work Available at: https://gitcoin.co/explorer

**Comment by vs77bb** [comment]
@lazaridiscom Glad to see you taking this on! This is definitely an experiment so we'll be excited to hear feedback from both you and @danfinlay as things progress 🙂

**Comment by lazaridiscom** [comment]
@vs77bb - Really? You (the gitcoin staff) should really not comment on issues subjecting the bounties. Stay neutral, please(!). So, lets try this "deleted comment" now. (you may want to file a "provide issue comment guidelines for gitcoin stuff", with the main rule "stay out of issues talk")

**Comment by danfinlay** [comment]
@lazaridiscom Whoah, what's wrong here? The Gitcoin staff posted this bounty, not us, so this is very much their experiment as much as it is ours. What could possibly be wrong with him chiming in with support here?

I'll add here that @lazaridiscom and I have talked a bit via email, and I agreed that since this issue is underpriced, it may be fair to award it for sufficient documenting/planning with potential proposed architecture. It probably would be most fair to clearly define that task if it's going to be treated as the bounty subject.

**Comment by lazaridiscom** [comment]
@danfinlay , I "fight" with gitcoin and its staff since days, mainly to reduce the side-noise within issues. I do this in the role of a "user" or "customer", so I possibly sound sometimes a bit itchy (as usual users and customers do).

> The Gitcoin staff posted this bounty, not us, so this is very much their experiment as much as it is ours.

Oh...!

Well, I'm human, and thus I've the right to make an error, right?

Boss! Sorry! (@vs77bb )

**Comment by danfinlay** [comment]
> I "fight" with gitcoin and its staff since days, mainly to reduce the side-noise within issues. I do this in the role of a "user" or "customer", so I possibly sound sometimes a bit itchy

As long as you're just trying to improve the product, and they understand the spirit, that's fine with me. It looked explosive without that context.

**Comment by lazaridiscom** [comment]
> As long as you're just trying to improve the product,

Of course. I like the product (even placed a bounty myself), and want it to become better.

> and they understand the spirit

"explosions" happened a few times within the gitcoin project (with parentheses explaining things calmly). They understand (and usually react to the feedback).

> It looked explosive without that context.

You're totally right (just went through). I may add that this "delete messages" thing is "under discussion" currently, too (e.g. deleting old message of the bot, or deleting unneeded messages of gitcoin staf). Without context, this "So, lets try this "deleted comment" now." of mine sounds even rude.

**Comment by lazaridiscom** [comment]
@danfinlay , can you please try to split this document into two major sections: core-library, ui-library. Those should be separated strictly (at least its the theoretical goal). If I reach implementation phase, then I will provide only the core.

@vs77bb, It looks like I'm throwing rocks here. Please confirm that you know its just pop-corn.

**Comment by owocki** [comment]
hey @lazaridiscom hit me up on slack when you get a sec tomorrow

**Comment by lazaridiscom** [comment]
@owocki , tomorrow? Sheduling is a totally creativity killer, especially during harder code-dives. I tried just now, only for Slack telling me you're on "Do Not Disturb" (instead of simply indication this by a nice glowing red dot). Now I have again enough of Slack for the next month or so.

Sent you and invite @ Hangouts. Telegram is nice for focused chat, too. Although you could simply send me an email, too.

**Comment by lazaridiscom** [comment]
@danfinlay , as usual, i delete my off-topic messages after some days, so don't wonder if it looks like you talk to yourself.

I have additionally "Stopped Work" on this Bounty (https://gitcoin.co/funding/details?url=https://github.com/MetaMask/metamask-extension/issues/3249).

I am still on this task, and I have the first estimation for you:

* 30 to 60 days.

The main problem is: the involved subsystems, the organizational parameters (see email).

**Comment by gitcoinbot** [comment]
__Work has been started on the 1.1 ETH (614.3 USD @ $558.46/ETH) funding by__: 
 1. [@skip-carlson](https://gitcoin.co/profile/skip-carlson) 

 __Please work together__ and coordinate delivery of the issue scope. Gitcoin doesn't know enough about everyones skillsets / free time to say who should work on what, but we trust that the community is smart and well-intentioned enough to work together.  As a general rule; if you start work first, youll be at the top of the above list ^^, and should have 'dibs' as long as you follow through. 

 On the above list? Please leave a comment to let the funder (@owocki) and the other parties involved what you're working, with respect to this issue and your plans to resolve it.  If you don't leave a comment, the funder may expire your submission at their discretion. 

 * Learn more [on the gitcoin issue page](https://gitcoin.co/issue/MetaMask/metamask-extension/3249)
 * Questions? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $3686.11 more Funded OSS Work Available at: https://gitcoin.co/explorer

**Comment by gitcoinbot** [comment]
__Work has been started on the 1.1 ETH (441.26 USD @ $401.14/ETH) funding by__: 
 1. [@Omyreon33](https://gitcoin.co/profile/Omyreon33) 

 __Please work together__ and coordinate delivery of the issue scope. Gitcoin doesn't know enough about everyones skillsets / free time to say who should work on what, but we trust that the community is smart and well-intentioned enough to work together.  As a general rule; if you start work first, youll be at the top of the above list ^^, and should have 'dibs' as long as you follow through. 

 On the above list? Please leave a comment to let the funder (@owocki) and the other parties involved what you're working, with respect to this issue and your plans to resolve it.  If you don't leave a comment, the funder may expire your submission at their discretion. 

 * Learn more [on the gitcoin issue page](https://gitcoin.co/issue/MetaMask/metamask-extension/3249)
 * Questions? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $2392.03 more Funded OSS Work Available at: https://gitcoin.co/explorer

**Comment by mbeacom** [comment]
__This issue now has a funding of 1.1 ETH (419.14 USD @ $381.04/ETH) attached to it.__

 * If you would like to work on this issue you can claim it [here](http://localhost:8000/issue/MetaMask/metamask-extension/3249).
 * If you've completed this issue and want to claim the bounty you can do so [here](http://localhost:8000/issue/MetaMask/metamask-extension/3249)
 * Questions? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $453.43 more Funded OSS Work Available at: https://gitcoin.co/explorer

**Comment by vs77bb** [comment]
Hi @omyreaon33 can you update with progress here? If not, Gitcoin Bot will likely open this one back up to the community. Hope you're having a good weekend!

**Comment by gitcoinbot** [comment]
@Omyreon33 are you still working on this issue?

**Comment by gitcoinbot** [comment]
@Omyreon33 are you still working on this issue?

**Comment by vs77bb** [comment]
This one has been opened back up on Gitcoin due to inactivity. Please take a look [here](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42) if you're interested in claiming! 

Apologies for the many comments from Gitcoin Bot all, we're cleaning this up. cc @owocki

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

__Work has been started on the 1.1 ETH (751.66 USD @ $683.33/ETH) funding by__: 
 1. [@St-Matthew](https://gitcoin.co/profile/St-Matthew)  

 On the above list? Please leave a comment to let the funder (@owocki) know your  plan. If you don't leave a comment, the funder may expire your submission at their discretion. 

 * Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42)
 * Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $8,334.69 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by gitcoinbot** [comment]
@St-Matthew Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] auto removal (6 days)

**Comment by gitcoinbot** [comment]
@St-Matthew Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] auto removal (6 days)

**Comment by gitcoinbot** [comment]
@St-Matthew Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] auto removal (6 days)

**Comment by gitcoinbot** [comment]
@St-Matthew Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=1) | [3 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=3) | [5 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=5) | [10 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=10) | [100 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=100)

**Comment by gitcoinbot** [comment]
@St-Matthew Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=1) | [3 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=3) | [5 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=5) | [10 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=10) | [100 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=100)

**Comment by St-Matthew** [comment]
I am not. This was selected by accident while testing the iOS app.

**Comment by weberV** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 1.1 ETH (761.31 USD @ $692.1/ETH) attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](http://localhost:8000/issue/MetaMask/metamask-extension/3249/42).
 * Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $823.60 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

__Work has been started__.

 1. [@rav8815](https://gitcoin.co/profile/rav8815) 

 has committed to working on this project to be completed 2 weeks, 2 days ago.


@owocki, __please see the below comments / questions regarding approach for this ticket from the bounty hunter(s):__

**Comment by gitcoinbot** [comment]
@rav8815 Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=1) | [3 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=3) | [5 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=5) | [10 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=10) | [100 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=100)

**Comment by gitcoinbot** [comment]
@thjnhcolag Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=1) | [3 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=3) | [5 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=5) | [10 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=10) | [100 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=100)

**Comment by gitcoinbot** [comment]
@thjnhcolag Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=1) | [3 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=3) | [5 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=5) | [10 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=10) | [100 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=100)

**Comment by gitcoinbot** [comment]
@thjnhcolag Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=1) | [3 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=3) | [5 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=5) | [10 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=10) | [100 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=100)

**Comment by ceresstation** [comment]
Hi @thjnhcolag unfortunately we'll need to give this issue back to the crowd as we haven't heard any updates from you. If you think we've made a mistake please let us know!

**Comment by lazaridiscom** [comment]
@ceresstation , this issue here is superseded by the (mostly architectural, Ledger focused, but of general validity) #4060 (and interlinked issues), and additionally by the (independent of #4060, Trezor focused, again general validity) practical work within #4625.

@danfinlay , should be possibly closed.

**Comment by ceresstation** [comment]
@lazaridiscom thanks for clarifying that

@danfinlay I see you're marked as the funder, would you be willing to close this bounty and reopen it on #4060? Much appreciated!

**Comment by gitcoinbot** [comment]
@thjnhcolag Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=1) | [3 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=3) | [5 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=5) | [10 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=10) | [100 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=100)

**Comment by danfinlay** [comment]
My opinion on this issue has returned to its original state: This is too large an issue to entrust to an external contributor or bounty hunter, and so the MetaMask team will be focusing on it internally, and removing the bounty from this issue.

The problem is mostly that tightly integrated refactors benefit from rapid communication, and we tend to have dramatically stronger/faster communication within the team than without. We'll continue to work towards improving our outwards-facing communication, but until all the required channels are in place, it doesn't make sense to encourage contributors like @ceresstation to pursue these kinds of tasks.

**Comment by lazaridiscom** [comment]
>  This is too large an issue to entrust to an external contributor

Larger issues can be broken down, making them manageable, as it was done here:

#3427, #3631, #4060 and https://github.com/lazaridiscom/mm-vault

As for "trust": sometimes you have to trust other contributors, and let them do their stuff.

> The problem is mostly that tightly integrated refactors benefit from rapid communication, and we tend to have dramatically stronger/faster communication within the team than without.

Refactoring/Redesign does not depend on direct/synchronous communication.

Larger Scale Refactoring/Redesign must be processed within mainly one brain anyway ( = zero communication barriers).

> We'll continue to work towards improving our outwards-facing communication, but until all the required channels are in place, it doesn't make sense to encourage contributors like @ceresstation to pursue these kinds of tasks.

Rearranging the project is trivial. This is not something that can be delayed for another year. Opening the project is a must. All resources are there. and them main points are:

* The relevant summaries of internal team discussions must be published as summaries on project resources (Issues etc.)
* Provide real milestones and road-maps
* No work starts without an Issue being filed first
* Work is done in an observable manner (e.g. no single-chunk large commits, but incremental smaller commits)
* Larger plans ar broken down into smaller ones (which can take some time, or it can be a task in itself).
* ...

-

Having said the above: this issue here should anyway be closed, unlabeled ("bounty-worthy") and closed on gitcoin.

And once you realize that "things are not getting better, despite MetaMask getting more staff", remember this on:

* Instead of throwing 5 more people on development, throw one person on redesigning code/process, thus the existed people become more productive.

Fix the Algorithm, instead of brute-force adding computing power.

This is... Standard Procedure!

(I'm most possibly available, so, whenever you decide to execute an combined redesign/refactioring etc., just let me know)

**Comment by gitcoinbot** [comment]
@thjnhcolag Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=1) | [3 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=3) | [5 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=5) | [10 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=10) | [100 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=100)

**Comment by danfinlay** [comment]
Closing as it has been replaced by more specific issues.

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__The funding of 1.1 ETH (513.11 USD @ $466.46/ETH) attached to this issue has been approved & issued.__  

 * Learn more at [on the Gitcoin Issue Details page](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42)
 * Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $18,307.10 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by gitcoinbot** [comment]
@thjnhcolag Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=1) | [3 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=3) | [5 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=5) | [10 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=10) | [100 days](https://gitcoin.co/issue/MetaMask/metamask-extension/3249/42?snooze=100)
